File: language-lab/elixir/runtime.exs
defmodule TinyP2P.LanguageLab.Runtime do
  alias TinyP2P.LanguageLab.{Atom, Kernel, Node}

  @bound 64

  @spec cycle(Node.t(), [binary()], non_neg_integer(), [binary()], non_neg_integer()) :: Node.t()
  def cycle(node, inbox, now_ms, shipped \\ [], bound \\ @bound) do
    node =
      Enum.reduce(inbox, node, fn data, acc ->
        {next, _owner} = Kernel.admit(acc, data)
        next
      end)

    Kernel.turn(node, now_ms, shipped, bound)
  end

  @spec outbox(Node.t()) :: [TinyP2P.LanguageLab.Row.t()]
  def outbox(node) do
    Kernel.watched(node, "send", "outbox") ++ Kernel.watched(node, "ship", "outbox")
  end

  @spec pump(Node.t(), function(), function(), MapSet.t(binary()), map()) ::
          {MapSet.t(binary()), map()}
  def pump(node, route, deliver, shipped, sent \\ %{}) do
    grouped =
      node
      |> outbox()
      |> Enum.reject(&MapSet.member?(shipped, &1.owner))
      |> Enum.group_by(& &1.owner, & &1.atom)
      |> Enum.sort_by(fn {owner, _atoms} -> owner end)

    Enum.reduce(grouped, {MapSet.new(), sent}, fn {owner, atoms}, {fired, sent_acc} ->
      cid = atoms |> hd() |> Map.fetch!(:target) |> target_low()

      case route.(cid) do
        nil ->
          {fired, sent_acc}

        {address, secret} ->
          seen = Map.get(sent_acc, cid, MapSet.new())
          candidates = shipment_candidates(atoms, node.durable, seen)
          inners = Enum.map(candidates, &elem(&1, 0))
          keys = Enum.map(candidates, &elem(&1, 1))

          delivered =
            if inners == [] do
              0
            else
              inners
              |> then(&deliver.(cid, address, secret, &1))
              |> max(0)
              |> min(length(inners))
            end

          seen =
            keys
            |> Enum.take(delivered)
            |> Enum.reject(&is_nil/1)
            |> Enum.reduce(seen, &MapSet.put(&2, &1))

          {MapSet.put(fired, owner), Map.put(sent_acc, cid, seen)}
      end
    end)
  end

  @spec wire_message(0..255, binary()) :: binary()
  def wire_message(kind, body) when kind in 0..255 and is_binary(body) do
    payload = <<kind, body::binary>>

    if byte_size(payload) >= 4_294_967_296 do
      raise ArgumentError, "wire payload must be shorter than 2^32 bytes"
    end

    <<byte_size(payload)::unsigned-big-32, payload::binary>>
  end

  defp shipment_candidates(atoms, durable, seen) do
    atoms
    |> Enum.sort_by(fn atom -> {atom.role, atom.value || <<>>} end)
    |> Enum.flat_map(fn
      %Atom{role: "send", value: value} when is_binary(value) ->
        [{value, nil}]

      %Atom{role: "ship", value: value} when is_binary(value) ->
        case Kernel.unframe(value) do
          {:ok, fact_ids} ->
            for fact_id <- fact_ids,
                Map.has_key?(durable, fact_id),
                not MapSet.member?(seen, fact_id) do
              {Map.fetch!(durable, fact_id), fact_id}
            end

          {:error, reason} ->
            raise ArgumentError, "invalid ship id framing: #{inspect(reason)}"
        end

      _other ->
        []
    end)
  end

  defp target_low({:exact, value}), do: value
  defp target_low({:range, low, _high}), do: low
end

defmodule TinyP2P.LanguageLab.WireDecoder do
  alias TinyP2P.LanguageLab.WireDecoder

  defstruct buffer: <<>>

  @type t :: %__MODULE__{buffer: binary()}

  @spec feed(t(), binary()) :: {[{0..255, binary()}], t()}
  def feed(%__MODULE__{} = decoder, data) when is_binary(data) do
    {messages, buffer} = decode(decoder.buffer <> data, [])
    {messages, %WireDecoder{buffer: buffer}}
  end

  defp decode(<<size::unsigned-big-32, rest::binary>>, messages)
       when byte_size(rest) >= size do
    <<payload::binary-size(size), tail::binary>> = rest

    case payload do
      <<>> -> decode(tail, messages)
      <<kind, body::binary>> -> decode(tail, [{kind, body} | messages])
    end
  end

  defp decode(buffer, messages), do: {Enum.reverse(messages), buffer}
end

defmodule TinyP2P.LanguageLab.OutLink do
  alias TinyP2P.LanguageLab.Runtime

  @enforce_keys [:capacity, :chunks]
  defstruct [:capacity, :chunks, head_offset: 0, pending: 0]

  @type t :: %__MODULE__{
          capacity: non_neg_integer(),
          chunks: :queue.queue(binary()),
          head_offset: non_neg_integer(),
          pending: non_neg_integer()
        }

  @spec new(non_neg_integer()) :: t()
  def new(capacity) when is_integer(capacity) and capacity >= 0 do
    %__MODULE__{capacity: capacity, chunks: :queue.new()}
  end

  @spec enqueue(t(), 0..255, binary()) :: {:ok, t()} | {:error, :full, t()}
  def enqueue(%__MODULE__{} = link, kind, body) do
    message = Runtime.wire_message(kind, body)

    if link.pending + byte_size(message) > link.capacity do
      {:error, :full, link}
    else
      {:ok,
       %{
         link
         | chunks: :queue.in(message, link.chunks),
           pending: link.pending + byte_size(message)
       }}
    end
  end

  @spec take(t(), integer()) :: {binary(), t()}
  def take(%__MODULE__{} = link, size) when is_integer(size) and size <= 0, do: {<<>>, link}

  def take(%__MODULE__{} = link, size) when is_integer(size) do
    wanted = min(size, link.pending)

    {chunks, head_offset, pieces, consumed} =
      take_chunks(link.chunks, link.head_offset, wanted, [], 0)

    data = pieces |> Enum.reverse() |> IO.iodata_to_binary()

    {data,
     %{
       link
       | chunks: chunks,
         head_offset: head_offset,
         pending: link.pending - consumed
     }}
  end

  defp take_chunks(chunks, head_offset, 0, pieces, consumed) do
    {chunks, head_offset, pieces, consumed}
  end

  defp take_chunks(chunks, head_offset, wanted, pieces, consumed) do
    case :queue.out(chunks) do
      {:empty, _chunks} ->
        {chunks, 0, pieces, consumed}

      {{:value, chunk}, tail} ->
        available = byte_size(chunk) - head_offset
        amount = min(available, wanted)
        piece = binary_part(chunk, head_offset, amount)

        if amount == available do
          take_chunks(tail, 0, wanted - amount, [piece | pieces], consumed + amount)
        else
          {chunks, head_offset + amount, [piece | pieces], consumed + amount}
        end
    end
  end
end

File: docs/language-lab/rust/src/runtime.rs
use crate::kernel::{CodecError, FactId, Node, Owner, Root, Row, unframe};
use std::collections::{BTreeMap, HashMap, HashSet};

pub const BOUND: usize = 64;

pub fn cycle<R: Root>(
    node: &mut Node<R>,
    inbox: &[Vec<u8>],
    now_ms: u64,
    shipped: &[FactId],
    bound: usize,
) {
    for data in inbox {
        node.admit(data);
    }
    node.turn(Some(now_ms), shipped, bound);
}

pub fn outbox<R: Root>(node: &Node<R>) -> Vec<Row> {
    let mut rows = node.watched(b"send", b"outbox");
    rows.extend(node.watched(b"ship", b"outbox"));
    rows
}

pub type Sent = HashMap<Vec<u8>, HashSet<FactId>>;

pub fn pump<R, Route, Deliver>(
    node: &Node<R>,
    mut route: Route,
    mut deliver: Deliver,
    shipped: &HashSet<FactId>,
    mut sent: Option<&mut Sent>,
) -> Result<HashSet<FactId>, CodecError>
where
    R: Root,
    Route: FnMut(&[u8]) -> Option<(Vec<u8>, Option<Vec<u8>>)>,
    Deliver: FnMut(&[u8], &[u8], Option<&[u8]>, &[Vec<u8>]) -> usize,
{
    let mut grouped = BTreeMap::<FactId, Vec<_>>::new();
    for row in outbox(node) {
        if let Owner::Fact(owner) = row.owner
            && !shipped.contains(&owner)
        {
            grouped.entry(owner).or_default().push(row.atom);
        }
    }
    let mut fired = HashSet::new();
    for (owner, mut atoms) in grouped {
        let Some(cid) = atoms
            .first()
            .and_then(|atom| atom.target.low())
            .map(<[u8]>::to_vec)
        else {
            continue;
        };
        let Some((address, secret)) = route(&cid) else {
            continue;
        };
        atoms.sort_by(|left, right| {
            (&left.role, left.value.as_deref().unwrap_or_default())
                .cmp(&(&right.role, right.value.as_deref().unwrap_or_default()))
        });
        let mut inners = Vec::new();
        let mut keys = Vec::new();
        let mut temporary_seen = HashSet::new();
        let seen = if let Some(all_sent) = sent.as_deref_mut() {
            all_sent.entry(cid.clone()).or_default()
        } else {
            &mut temporary_seen
        };
        for atom in atoms {
            match (atom.role.as_slice(), atom.value) {
                (b"send", Some(value)) => {
                    inners.push(value);
                    keys.push(None);
                }
                (b"ship", Some(value)) => {
                    for fact_key in unframe(&value)? {
                        let Ok(fact_id) = FactId::try_from(fact_key.as_slice()) else {
                            continue;
                        };
                        if let Some(data) = node.durable.get(&fact_id)
                            && !seen.contains(&fact_id)
                        {
                            inners.push(data.clone());
                            keys.push(Some(fact_id));
                        }
                    }
                }
                _ => {}
            }
        }
        if !inners.is_empty() {
            let delivered = deliver(&cid, &address, secret.as_deref(), &inners).min(inners.len());
            for fact_id in keys.into_iter().take(delivered).flatten() {
                seen.insert(fact_id);
            }
        }
        fired.insert(owner);
    }
    Ok(fired)
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum WireError {
    MessageTooLarge,
}

pub fn wire_message(kind: u8, body: &[u8]) -> Result<Vec<u8>, WireError> {
    let size = body
        .len()
        .checked_add(1)
        .and_then(|length| u32::try_from(length).ok())
        .ok_or(WireError::MessageTooLarge)?;
    let mut message = Vec::with_capacity(4 + size as usize);
    message.extend_from_slice(&size.to_be_bytes());
    message.push(kind);
    message.extend_from_slice(body);
    Ok(message)
}

#[derive(Clone, Debug, Default)]
pub struct WireDecoder {
    buffer: Vec<u8>,
}

impl WireDecoder {
    pub fn feed(&mut self, data: &[u8]) -> Vec<(u8, Vec<u8>)> {
        self.buffer.extend_from_slice(data);
        let mut messages = Vec::new();
        let mut offset = 0;
        while self.buffer.len().saturating_sub(offset) >= 4 {
            let size = u32::from_be_bytes(
                self.buffer[offset..offset + 4]
                    .try_into()
                    .expect("four-byte slice"),
            ) as usize;
            let Some(end) = offset
                .checked_add(4)
                .and_then(|start| start.checked_add(size))
            else {
                break;
            };
            if end > self.buffer.len() {
                break;
            }
            if size > 0 {
                messages.push((
                    self.buffer[offset + 4],
                    self.buffer[offset + 5..end].to_vec(),
                ));
            }
            offset = end;
        }
        if offset > 0 {
            self.buffer.drain(..offset);
        }
        messages
    }

    pub fn buffered_len(&self) -> usize {
        self.buffer.len()
    }
}

#[derive(Clone, Debug)]
pub struct OutLink {
    capacity: usize,
    buffer: Vec<u8>,
    offset: usize,
}

impl OutLink {
    pub fn new(capacity: usize) -> Self {
        Self {
            capacity,
            buffer: Vec::new(),
            offset: 0,
        }
    }

    pub fn pending(&self) -> usize {
        self.buffer.len() - self.offset
    }

    pub fn enqueue(&mut self, kind: u8, body: &[u8]) -> Result<bool, WireError> {
        let message = wire_message(kind, body)?;
        if self
            .pending()
            .checked_add(message.len())
            .is_none_or(|size| size > self.capacity)
        {
            return Ok(false);
        }
        self.buffer.extend(message);
        Ok(true)
    }

    pub fn take(&mut self, size: usize) -> Vec<u8> {
        let end = self.buffer.len().min(self.offset.saturating_add(size));
        let data = self.buffer[self.offset..end].to_vec();
        self.offset = end;
        if self.offset == self.buffer.len() {
            self.buffer.clear();
            self.offset = 0;
        } else if self.offset > 1 << 20 {
            self.buffer.drain(..self.offset);
            self.offset = 0;
        }
        data
    }
}
